### ISSUE TITLE
Error: Incorrect user or password

### ISSUE THREAD

**Issue by jacargentina** [keystonejs/keystone#1697]
I'm getting this stack:

````
------------------------------------------------
Error handling 500 (error):
[TypeError: res.err is not a function]

------------------------------------------------

Error thrown for request: /
Error: Incorrect user or password
    at /home/javiercastro/javascript/website/node_modules/keystone/lib/session.js:113:19
    at Query.<anonymous> (/home/javiercastro/javascript/website/node_modules/keystone/node_modules/mongoose/lib/query.js:2119:28)
    at /home/javiercastro/javascript/website/node_modules/keystone/node_modules/mongoose/node_modules/kareem/index.js:177:19
    at /home/javiercastro/javascript/website/node_modules/keystone/node_modules/mongoose/node_modules/kareem/index.js:109:16
    at doNTCallback0 (node.js:407:9)
    at process._tickCallback (node.js:336:13)
GET / 500 67.312 ms
````

I think it may be related to this: after many months, i've opened my Firefox browser (normally i use Chrome every day) and pointed to my dev keystone site. Then i get the stack.

I assume it is related to the old session cookie stored on Firefox which is now invalid (may be i've changed the secret after my last Firefox use on this site)

Which would be the way to avoid/correct this?

**Comment by JedWatson** [comment]
Unfortunately I'll need more info to help with this one; whatever is being passed as the `onErr` argument to `doSignin` in `lib/session.js` is referring to `res.err` which doesn't exist.

Can you track down where that's being called from? Is it something in your dev site, or part of Keystone itself? also which version are you running? (0.3.x / 0.4 alpha / etc)

**Comment by LorbusChris** [comment]
I can confirm this one, although under slightly different circumstances and without any mention of `res.err`:
I'm currently using Firefox on a Fedora 22 Workstation. As there's no MongoDB support yet, I went with mockgoose, which worked just fine the first time starting keystone. However, when stopping and restarting, I got the error output below. I deleted the cookies before the last 2 lines and it went back to working. It also didn't happen whenever I had NOT signed into Keystone before restarting. So its certainly a cookie issue that's occurring, even though I never changed the secret. 

```
------------------------------------------------
KeystoneJS Started:
Keystone Test is ready on port 3000
------------------------------------------------

Error thrown for request: /keystone/
Error: Incorrect user or password
    at /home/cglombek/js/node_modules/keystone/lib/session.js:113:19
    at Query.<anonymous> (/home/cglombek/js/node_modules/mongoose/lib/query.js:2119:28)
    at /home/cglombek/js/node_modules/kareem/index.js:177:19
    at /home/cglombek/js/node_modules/kareem/index.js:109:16
    at doNTCallback0 (node.js:408:9)
    at process._tickCallback (node.js:337:13)
GET /keystone/ 500 66.474 ms
Error thrown for request: /nothinghere
Error: Incorrect user or password
    at /home/cglombek/js/node_modules/keystone/lib/session.js:113:19
    at Query.<anonymous> (/home/cglombek/js/node_modules/mongoose/lib/query.js:2119:28)
    at /home/cglombek/js/node_modules/kareem/index.js:177:19
    at /home/cglombek/js/node_modules/kareem/index.js:109:16
    at doNTCallback0 (node.js:408:9)
    at process._tickCallback (node.js:337:13)
GET /keystone/ 302 6.708 ms
GET /keystone/signin 200 125.260 ms
```

¡Saludos Che!

**Comment by w01fgang** [comment]
I get this error when the DB of the site is changed and the username doesn't changed.
For override this error I delete cookies of that site.

**Comment by LorbusChris** [comment]
For me, Mockgoose would set up a new in-memory db each time starting the server, so that would totally make sense.

**Comment by JedWatson** [comment]
This was a pretty subtle fix, based on @w01fgang's work in #1697. The key was both to remove the cookie and _not_ pass the error back to express; with this change middleware that checks for a valid session bounces you correctly to the signing screen.
